### PR TITLE
Handle create () in cost planner

### PR DESCRIPTION
**Pull request by pontusmelke**
Initial support for `create (a :L {prop: 42})` in cost planner.

### CODE REVIEW

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CreateAcceptanceTest.scala [hunk 1]
```diff
@@ -19,32 +19,81 @@
...
+    assertStats(result, nodesCreated = 1, propertiesSet = 1)
+    // then
+    result.toList should equal(List(Map("p" -> "foo")))
+  }
 
   test("using an undirected relationship pattern should fail on create") {
-    evaluating {
-      executeScalar[Relationship]("create (a {id: 2})-[r:KNOWS]-(b {id: 1}) RETURN r")
-    }  should produce[SyntaxException]
+      intercept[SyntaxException](executeScalar[Relationship]("create (a {id: 2})-[r:KNOWS]-(b {id: 1}) RETURN r"))
```
Comment (Mats-SX) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/CreateAcceptanceTest.scala:82]: I think a nicer syntax is `a [SyntaxException] should be thrownBy(<call>)`, but this works too.

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/ReturnAcceptanceTest.scala [hunk 2]
```diff
@@ -313,7 +313,7 @@ order by a.age""").toList)
   }
 
   test("renaming in multiple steps should still work") {
-    val result = executeWithRulePlanner(
+    val result = executeWithCostPlannerOnly(
```
Comment (Mats-SX) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/ReturnAcceptanceTest.scala:316]: Shouldn't this be `updateWithBothPlanners`?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/ast/convert/plannerQuery/PlannerQueryBuilder.scala [hunk 3]
```diff
@@ -29,8 +29,11 @@ case class PlannerQueryBuilder(private val q: PlannerQuery, returns: Seq[IdName]
 
   def withReturns(returns: Seq[IdName]): PlannerQueryBuilder = copy(returns = returns)
 
-  def updateGraph(f: QueryGraph => QueryGraph): PlannerQueryBuilder =
-    copy(q = q.updateTailOrSelf(_.updateGraph(f)))
+  def updateQueryGraph(f: QueryGraph => QueryGraph): PlannerQueryBuilder =
+    copy(q = q.updateTailOrSelf(_.updateQueryGraph(f)))
+
+  def updateUpdateGraph(f: UpdateGraph => UpdateGraph): PlannerQueryBuilder =
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/ast/convert/plannerQuery/PlannerQueryBuilder.scala]: This is a bit trippy... How about calling these `amendQueryGraph` and `amendUpdateGraph` instead?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/EntityProducerFactory.scala [hunk 4]
```diff
@@ -21,7 +21,7 @@ package org.neo4j.cypher.internal.compiler.v3_0.commands
 
 import org.neo4j.cypher.internal.compiler.v3_0._
 import org.neo4j.cypher.internal.compiler.v3_0.executionplan.builders.GetGraphElements
-import org.neo4j.cypher.internal.compiler.v3_0.mutation.GraphElementPropertyFunctions
+import org.neo4j.cypher.internal.compiler.v3_0.mutation.{makeValueNeoSafe, GraphElementPropertyFunctions}
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/EntityProducerFactory.scala]: Remove

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/indexQuery.scala [hunk 5]
```diff
@@ -22,7 +22,7 @@ package org.neo4j.cypher.internal.compiler.v3_0.commands
 import org.neo4j.cypher.internal.compiler.v3_0.ExecutionContext
 import org.neo4j.cypher.internal.compiler.v3_0.commands.expressions.{Expression, InequalitySeekRangeExpression, PrefixSeekRangeExpression}
 import org.neo4j.cypher.internal.compiler.v3_0.helpers.IsCollection
-import org.neo4j.cypher.internal.compiler.v3_0.mutation.GraphElementPropertyFunctions
+import org.neo4j.cypher.internal.compiler.v3_0.mutation.{GraphElementPropertyFunctions, makeValueNeoSafe}
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/indexQuery.scala]: Remove

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/EntityProducerFactory.scala (hunk 4 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/commands/EntityProducerFactory.scala]: Nevermind, I saw below.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/AllNodesScanPipe.scala [hunk 6]
```diff
@@ -41,7 +41,7 @@ case class AllNodesScanPipe(ident: String)(val estimatedCardinality: Option[Doub
 
   override def monitor = pipeMonitor
 
-  override def localEffects: Effects = Effects(ReadsAllNodes)
+  override def localEffects: Effects = Effects(ReadsAllNodes).asLeafEffects
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/AllNodesScanPipe.scala:44]: We decided to move away from effects on pipes, right? If that's correct, remove this.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/ApplyPipe.scala [hunk 7]
```diff
@@ -52,3 +52,5 @@ case class ApplyPipe(source: Pipe, inner: Pipe)(val estimatedCardinality: Option
 
   def withEstimatedCardinality(estimated: Double) = copy()(Some(estimated))
 }
+
+
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/ApplyPipe.scala]: noise

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/EmptyResultPipe.scala [hunk 8]
```diff
@@ -46,4 +47,13 @@ case class EmptyResultPipe(source: Pipe)(implicit pipeMonitor: PipeMonitor) exte
     val (source :: Nil) = sources
     copy(source = source)
   }
+
+  def estimatedCardinality: Option[Double] = Some(0.0)
+
+  override def planDescriptionWithoutCardinality: InternalPlanDescription = new SingleRowPlanDescription(this.id, Seq.empty, identifiers)
+
+  override def withEstimatedCardinality(estimated: Double): Pipe with RonjaPipe = {
+    //TODO should enforce estimated == 0.0
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/EmptyResultPipe.scala:56]: Fix todo?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/ExecuteUpdateCommandsPipe.scala [hunk 9]
```diff
@@ -49,14 +48,13 @@ case class ExecuteUpdateCommandsPipe(source: Pipe, commands: Seq[UpdateAction])(
...
-                   singleCommand: Boolean): Iterator[ExecutionContext] = {
+                   state: QueryState): Iterator[ExecutionContext] = {
 
     val result: Iterator[ExecutionContext] = cmd.exec(ctx, state)
 
     cmd match {
-      case _:CreateNode if !singleCommand =>
-        singleOr(result, new ParameterWrongTypeException("If you create multiple elements, you can only create one of each."))
+      case _:CreateNode =>
+        singleOr(result, new CypherTypeException("If you want to create multiple nodes, please use UNWIND."))
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/ExecuteUpdateCommandsPipe.scala:57]: Perhaps a reference to an example in the manual could be in place here?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/CostBasedExecutablePlanBuilder.scala [hunk 10]
```diff
@@ -89,8 +89,8 @@ case class CostBasedExecutablePlanBuilder(monitors: Monitors,
 
     val pipeBuildContext = PipeExecutionBuilderContext(metrics.cardinality, semanticTable, plannerName)
 
-    //Check for notifications
-    checkForUnresolvedTokens(ast, semanticTable).foreach(notificationLogger += _)
+    //Check for unresolved tokens for read-only queries
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/CostBasedExecutablePlanBuilder.scala:92]: Why are we not doing this for mixed queries?

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/execution/PipeExecutionPlanBuilder.scala [hunk 11]
```diff
@@ -52,9 +53,8 @@ class PipeExecutionPlanBuilder(clock: Clock, monitors: Monitors) {
 
   def build(plan: LogicalPlan)(implicit context: PipeExecutionBuilderContext, planContext: PlanContext): PipeInfo = {
     implicit val table: SemanticTable = context.semanticTable
-    val updating = false
 
-    def buildPipe(plan: LogicalPlan): Pipe with RonjaPipe = {
+    def buildPipe(plan: LogicalPlan, updating: Boolean = false): Pipe with RonjaPipe = {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/execution/PipeExecutionPlanBuilder.scala]: Remove default value?

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/ReturnAcceptanceTest.scala (hunk 2 above)
Comment (pontusmelke) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/ReturnAcceptanceTest.scala:316]: The test does a lot of weird things with node ids, that makes the results between cost and rule not to be comparable

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/AllNodesScanPipe.scala (hunk 6 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/AllNodesScanPipe.scala:44]: We still need to implement effects as long as the rule planner is still around

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/CostBasedExecutablePlanBuilder.scala (hunk 10 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/CostBasedExecutablePlanBuilder.scala:92]: I though it was annoying to warn the user when doing `CREATE (m: Person)` when `Person` is not in the database. The user is about to create it and it felt noisy to throw a warning in this case. The other reason is that there are tests asserting that all the guides in the browser runs without giving notifications, and these guides typically start from an empty database and create nodes.  I am not sure what is the correct thing here.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/AllNodesScanPipe.scala (hunk 6 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/AllNodesScanPipe.scala:44]: That's true, but the rule planner doesn't use AllNodesScanPipe as far as I'm aware.

Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/AllNodesScanPipe.scala:44]: So you are saying we should implement it but make it return the wrong thing?

**Pull request merged**